sqlitecluster/SQLiteNode.cpp

```diff
@@ -1371,22 +1559,53 @@ bool SQLiteNode::update(uint64_t& nextActivity) {
                             SASSERTWARN(_currentCommand->transaction.empty());
                             SINFO("Starting processing command '" << _currentCommand->request.methodLine << "' ("
                                                                   << _currentCommand->id << ")");
-                            _processCommandWrapper(_db, _currentCommand);
+
+                            bool unlock = false;
+                            if (_currentCommand->processed >= MAX_PROCESS_TRIES) {
+                                SWARN("[concurrent] Command " << _currentCommand->id << " processed " << _currentCommand->processed
+                                      << " times, exceeded max of " << MAX_PROCESS_TRIES
+                                      << ", forcing synchronous commit.");
+                                unlock = true;
+                                SQLite::commitLock.lock();
+                            }
+                            bool needsCommit = _processCommandWrapper(_db, _currentCommand);
+
                             SASSERT(!_currentCommand->response.empty()); // Must set a response
 
                             // Anything to commit?
-                            if (_db.insideTransaction()) {
+                            if (needsCommit) {
+
+                                // We're about to send a transaction here, grab our commit mutex and send anything
+                                // outstanding. This remains locked until we've finished this transaction.
+                                SQLite::commitLock.lock();
```

Comment:
If we're doing a syncronous commit, we force that lock [here](https://github.com/Expensify/Bedrock/blob/b418206fb23c7fbb95b0d243264d65ae1a3e9be0/sqlitecluster/SQLiteNode.cpp#L1564), which locks from before `process()`.

Then, for any other commit, we lock [here](https://github.com/Expensify/Bedrock/blob/b418206fb23c7fbb95b0d243264d65ae1a3e9be0/sqlitecluster/SQLiteNode.cpp#L1576), and we hold this lock until we finish the commit.

Then [here](https://github.com/Expensify/Bedrock/blob/b418206fb23c7fbb95b0d243264d65ae1a3e9be0/sqlitecluster/SQLiteNode.cpp#L1612) we release the first of these locks, if it was held because it's a synchronous commit. In this case, we're still holding the lock once, and it will unlock when we've finished the distributed commit.

We unlock [here](https://github.com/Expensify/Bedrock/blob/b418206fb23c7fbb95b0d243264d65ae1a3e9be0/sqlitecluster/SQLiteNode.cpp#L1637) if we were going to force a synchronous commit, but there's nothing to commit.

Then we wait for the distributed transaction to complete and unlock [here](https://github.com/Expensify/Bedrock/blob/b418206fb23c7fbb95b0d243264d65ae1a3e9be0/sqlitecluster/SQLiteNode.cpp#L1301) if we had a conflict and will need to retry this transaction.

Or, finally, we unlock [here](https://github.com/Expensify/Bedrock/blob/b418206fb23c7fbb95b0d243264d65ae1a3e9be0/sqlitecluster/SQLiteNode.cpp#L1374) if the commit completed successfully.

This will probably all change as we refactor out `SQLiteCore` as well.